File: problems/886/solution.go
package main
func possibleBipartition(N int, dislikes [][]int) bool {
    graph, colors := make([][]int, N + 1), make(map[int]int)
    
    for _, edge := range dislikes {
        graph[edge[0]] = append(graph[edge[0]], edge[1])
        graph[edge[1]] = append(graph[edge[1]], edge[0])
    }
    
    for node := 1; node <= N; node++ {
        if !has(colors, node) && !bfs(graph, node, colors) {
            return false
        }
    }
    
    return true
}

func bfs(graph [][]int, node int, colors map[int]int) bool {
    q := []int{node}
    colors[node] = 1
    
    for len(q) > 0 {
        cur := q[0]
        q = q[1:]
        
        for _, adj := range graph[cur] {
            if !has(colors, adj) {
                q = append(q, adj)
                colors[adj] = colors[cur] ^ 1
            } else if colors[cur] == colors[adj] {
                return false
            }
        }
    }
    
    return true
}

func has(m map[int]int, n int) bool {
    _, ok := m[n]
    
    return ok
}

File: problems/886/solution.cpp
class Solution {
public:
    bool possibleBipartition(int n, vector<vector<int>>& dislikes) {
        unordered_map<int, unordered_set<int>> g;
        vector<int> colors(n + 1);
        
        for (const vector<int>& dislike: dislikes) {
            g[dislike[0]].insert(dislike[1]);
            g[dislike[1]].insert(dislike[0]);
        }
        
        for (int i = 1; i <= n; i++) {
            if (colors[i] != 0) {
                continue;
            }
            
            queue<int> q;
            
            colors[i] = 1;
            q.push(i);
            
            while (not empty(q)) {
                int node = q.front();
                q.pop();
                
                for (const int& next: g[node]) {
                    if (colors[next] != 0) {
                        if (colors[next] == colors[node]) {
                            return false;
                        }
                    } else {
                        q.push(next);
                        colors[next] -= colors[node];
                    }
                }
            }
        }
        
        return true;
    }
};